File: frontend/packages/web/src/configs/features.ts
import { ParseKeys } from 'i18next';

import { FeatureGroup } from '../enums/FeatureGroup';
import { RawFeature } from '../enums/RawFeature';
import { Feature } from '../interfaces/Feature';

export const RAW_FEATURES = Object.values(RawFeature);

export const UNSUPPORTED_RAW_FEATURES = import.meta.env.VITE_UNSUPPORTED_RAW_FEATURES?.split(',') ?? [];
export const UNUSED_RAW_FEATURES = [
  RawFeature.ANAG_SUBJECT_CATEGORY,
  RawFeature.PROP_REGISTRATION_OFFICES,
  RawFeature.COMMON_CITIES,
  RawFeature.ADMIN_SETTINGS,
  RawFeature.ECON_TAX_CREDIT_BASE,
  RawFeature.FCLT_SERVICE,
] as const;

export const FEATURES = RAW_FEATURES.filter((feature) => !(feature in UNUSED_RAW_FEATURES));

export const FEATURE_GROUP_NAMES: Record<FeatureGroup, ParseKeys> = {
  [FeatureGroup.Registry]: 'core.feature.registry.title',
  [FeatureGroup.RealEstate]: 'core.feature.real_estate.title',
  [FeatureGroup.Common]: 'core.feature.common.title',
  [FeatureGroup.AssetManagement]: 'core.feature.asset_management.title',
  [FeatureGroup.Repository]: 'core.feature.repository.title',
  [FeatureGroup.EnergyManagement]: 'core.feature.energy_management.title',
  [FeatureGroup.Maintenance]: 'core.feature.maintenance.title',
  [FeatureGroup.System]: 'core.feature.system.title',
} as const;

export const FEATURE_NAMES: Record<Feature, ParseKeys> = {
  [RawFeature.ANAG_SUBJECT_BASE]: 'core.feature.registry.subjects',
  [RawFeature.ANAG_SUBJECT_ORGUNIT]: 'core.feature.registry.org_units',
  [RawFeature.ASST_ESTATE_BASE]: 'core.feature.real_estate.estates',
  [RawFeature.ASST_ESTATE_CATALOGUE]: 'core.feature.real_estate.estates_catalogue',
  [RawFeature.ASST_ESTATEUNIT_BASE]: 'core.feature.real_estate.estate_units',
  [RawFeature.ASST_ESTATEUNIT_VARIATION]: 'core.feature.real_estate.estate_units_variations',
  [RawFeature.ASST_CADASTRALUNIT_BASE]: 'core.feature.real_estate.cadastral_units',
  [RawFeature.ASST_CATALOGUE_CONFIG]: 'core.feature.real_estate.catalogue_config',
  [RawFeature.ASST_CADASTRAL_LAND_CATEGORY]: 'core.feature.real_estate.cadastral_land_categories',
  [RawFeature.ASST_ASSET_TAX_BASE]: 'core.feature.real_estate.asset_taxes',
  [RawFeature.COMMON_VATRATES]: 'core.feature.common.vat_rates',
  [RawFeature.COMMON_INTERESTRATES]: 'core.feature.common.interest_rates',
  [RawFeature.COMMON_ACCOUNTINGITEMS]: 'core.feature.common.accounting_items',
  [RawFeature.COMMON_BILLITEMTYPES]: 'core.feature.common.bill_item_types',
  [RawFeature.COMMON_REVALUATIONDATA]: 'core.feature.common.revaluation_data',
  [RawFeature.COMMON_TAX_RATES]: 'core.feature.common.tax_rates',
  [RawFeature.COMMON_REPORT_GENERATORS]: 'core.feature.common.report_generators',
  [RawFeature.DOCUMENTS_BASE]: 'core.feature.repository.documents',
  [RawFeature.PROP_CONTRACT_BASE]: 'core.feature.asset_management.contracts',
  [RawFeature.PROP_CONTRACT_TYPES]: 'core.feature.asset_management.contract_types',
  [RawFeature.PROP_BILL_BASE]: 'core.feature.asset_management.bills',
  [RawFeature.PROP_ADMINISTRATION_BASE]: 'core.feature.asset_management.administration',
  [RawFeature.PROP_REGISTRY_COMMUNICATION]: 'core.feature.asset_management.registry_communications',
  [RawFeature.PROP_REGISTRATION_PAYMENT]: 'core.feature.asset_management.registration_payments',
  [RawFeature.NRGY_TYPE_BASE]: 'core.feature.energy_management.utility_types',
  [RawFeature.NRGY_SERVICE_BASE]: 'core.feature.energy_management.utility_services',
  [RawFeature.NRGY_COSTCHARGE_BASE]: 'core.feature.energy_management.cost_charges',
  [RawFeature.ADMIN_USERS_AND_GROUPS]: 'core.feature.system.permissions',
  [RawFeature.FCLT_ESTATE_UNIT_GROUP_BASE]: 'core.feature.maintenance.estate_unit_groups',
  [RawFeature.FCLT_CONTRACT_BASE]: 'core.feature.maintenance.facility_contracts',
  [RawFeature.FCLT_CONTRACT_TEMPLATES]: 'core.feature.maintenance.facility_contract_templates',
  [RawFeature.FCLT_CONTRACT_TYPES]: 'core.feature.maintenance.facility_contract_types',
  [RawFeature.FCLT_SLA_BASE]: 'core.feature.maintenance.slas',
  [RawFeature.FCLT_CONFIG]: 'core.feature.maintenance.config',
  [RawFeature.FCLT_PENALTY_BASE]: 'core.feature.maintenance.penalties',
  [RawFeature.FCLT_PRICE_LIST_BASE]: 'core.feature.maintenance.price_lists',
  [RawFeature.FCLT_PRICE_LIST_ARTICLES]: 'core.feature.maintenance.price_list_articles',
  [RawFeature.FCLT_TICKET]: 'core.feature.maintenance.tickets',
  [RawFeature.FCLT_SERVICE_CATEGORY]: 'core.feature.maintenance.service_categories',
  [RawFeature.ADMIN_AUDIT_LOG]: 'core.feature.system.audit_events',
  [RawFeature.ADMIN_CONFIG]: 'core.feature.system.configs',
} as const;

File: backend/src/Core/IAM/Features.cs
using System.Reflection;

namespace RealGimm.Core.IAM;

public static class Features
{
  public const string ANAG_SUBJECT_BASE = "121";
  public const string ANAG_SUBJECT_CATEGORY = "131";
  public const string ANAG_SUBJECT_ORGUNIT = "126";
  public const string ASST_ESTATE_BASE = "211";
  public const string ASST_ESTATE_CATALOGUE = "212";
  public const string ASST_ESTATEUNIT_BASE = "214";
  public const string ASST_ESTATEUNIT_VARIATION = "215";
  public const string ASST_CADASTRALUNIT_BASE = "217";
  public const string ASST_CATALOGUE_CONFIG = "230";
  public const string ASST_CADASTRAL_LAND_CATEGORY = "240";
  public const string ASST_ASSET_TAX_BASE = "250";
  public const string COMMON_VATRATES = "311";
  public const string COMMON_INTERESTRATES = "321";
  public const string COMMON_ACCOUNTINGITEMS = "322";
  public const string COMMON_BILLITEMTYPES = "323";
  public const string COMMON_REVALUATIONDATA = "324";
  public const string COMMON_CITIES = "331";
  public const string COMMON_TAX_RATES = "341";
  public const string COMMON_REPORT_GENERATORS = "351";
  public const string DOCUMENTS_BASE = "401";
  public const string PROP_CONTRACT_BASE = "501";
  public const string PROP_CONTRACT_TYPES = "511";
  public const string PROP_REGISTRATION_OFFICES = "512";
  public const string PROP_BILL_BASE = "521";
  public const string PROP_ADMINISTRATION_BASE = "531";
  public const string PROP_REGISTRY_COMMUNICATION = "541";
  public const string PROP_REGISTRATION_PAYMENT = "551";
  public const string NRGY_TYPE_BASE = "601";
  public const string NRGY_SERVICE_BASE = "611";
  public const string NRGY_COSTCHARGE_BASE = "621";
  public const string ECON_TAX_CREDIT_BASE = "701";
  public const string FCLT_ESTATE_UNIT_GROUP_BASE = "800";
  public const string FCLT_CONTRACT_BASE = "810";
  public const string FCLT_CONTRACT_TEMPLATES = "811";
  public const string FCLT_CONTRACT_TYPES = "812";
  public const string FCLT_SLA_BASE = "820";
  public const string FCLT_CONFIG = "830";
  public const string FCLT_PENALTY_BASE = "840";
  public const string FCLT_PRICE_LIST_BASE = "850";
  public const string FCLT_PRICE_LIST_ARTICLES = "851";
  public const string FCLT_TICKET = "860";
  public const string FCLT_SERVICE_CATEGORY = "865";
  public const string FCLT_SERVICE = "870";
  public const string ADMIN_SETTINGS = "911";
  public const string ADMIN_USERS_AND_GROUPS = "921";
  public const string ADMIN_AUDIT_LOG = "922";
  public const string ADMIN_CONFIG = "923";

  public static IEnumerable<string> UserFeatureList()
  {
    return typeof(Features)
        .GetFields(BindingFlags.Public | BindingFlags.Static | BindingFlags.FlattenHierarchy)
        .Where(f => f.IsLiteral && !f.IsInitOnly && f.FieldType == typeof(string) && !f.Name.StartsWith("ADMIN_"))
        .Select(f => (f.GetRawConstantValue() as string) ?? string.Empty);
  }

  public static IEnumerable<string> AdminFeatureList()
  {
    return typeof(Features)
        .GetFields(BindingFlags.Public | BindingFlags.Static | BindingFlags.FlattenHierarchy)
        .Where(f => f.IsLiteral && !f.IsInitOnly && f.FieldType == typeof(string) && f.Name.StartsWith("ADMIN_"))
        .Select(f => (f.GetRawConstantValue() as string) ?? string.Empty);
  }

  public static IEnumerable<string> AllFeatureList()
  {
    return typeof(Features)
        .GetFields(BindingFlags.Public | BindingFlags.Static | BindingFlags.FlattenHierarchy)
        .Where(f => f.IsLiteral && !f.IsInitOnly && f.FieldType == typeof(string))
        .Select(f => (f.GetRawConstantValue() as string) ?? string.Empty);
  }
}
